libbeat/keystore/file_keystore.go

```diff
@@ -0,0 +1,383 @@
+var version = []byte("v1")
+
+// FileKeystore Allows to store key / secrets pair securely into an encrypted local file.
+type FileKeystore struct {
+	sync.RWMutex
+	Path     string
+	secrets  map[string]serializableSecureString
+	dirty    bool
+	password *SecureString
+}
+
+// Allow the original SecureString type to be correctly serialized to json.
+type serializableSecureString struct {
+	*SecureString
+	Value []byte `json:"value"`
+}
+
+// NewFileKeystore returns an new File based keystore or an error, currently users cannot set their
+// own password on the keystore, the default password will be an empty string. When the keystore
+// is initialied the secrets are automatically loaded into memory.
+func NewFileKeystore(keystoreFile string) (Keystore, error) {
+	return NewFileKeystoreWithPassword(keystoreFile, NewSecureString([]byte("")))
+}
+
+// NewFileKeystoreWithPassword return a new File based keystore or an error, allow to define what
+// password to use to create the keystore.
+func NewFileKeystoreWithPassword(keystoreFile string, password *SecureString) (Keystore, error) {
+	keystore := FileKeystore{
+		Path:     keystoreFile,
+		dirty:    false,
+		password: password,
+		secrets:  make(map[string]serializableSecureString),
+	}
+
+	err := keystore.load()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &keystore, nil
+}
+
+// Retrieve return a SecureString instance that will contains both the key and the secret.
+func (k *FileKeystore) Retrieve(key string) (*SecureString, error) {
+	k.RLock()
+	defer k.RUnlock()
+
+	secret, ok := k.secrets[key]
+	if !ok {
+		return nil, ErrKeyDoesntExists
+	}
+	return NewSecureString(secret.Value), nil
+}
+
+// Store add the key pair to the secret store and mark the store as dirty.
+func (k *FileKeystore) Store(key string, value []byte) error {
+	k.Lock()
+	defer k.Unlock()
+
+	k.secrets[key] = serializableSecureString{Value: value}
+	k.dirty = true
+	return nil
+}
+
+// Delete an existing key from the store and mark the store as dirty.
+func (k *FileKeystore) Delete(key string) error {
+	k.Lock()
+	defer k.Unlock()
+
+	delete(k.secrets, key)
+	k.dirty = true
+	return nil
+}
+
+// Save persists the in memory data to disk if needed.
+func (k *FileKeystore) Save() error {
+	k.Lock()
+	defer k.Unlock()
+
+	if k.dirty == false {
+		return nil
+	}
+
+	temporaryPath := fmt.Sprintf("%s.tmp", k.Path)
+	if _, err := os.Stat(temporaryPath); os.IsExist(err) {
+		return fmt.Errorf("temporary keystore exist at '%s', delete it and try again", temporaryPath)
+	}
+
+	w := new(bytes.Buffer)
+	jsonEncoder := json.NewEncoder(w)
+	if err := jsonEncoder.Encode(k.secrets); err != nil {
+		return fmt.Errorf("cannot serialize the keystore before saving it to disk: %v", err)
+	}
+
+	encrypted, err := k.encrypt(w)
+	if err != nil {
+		return fmt.Errorf("cannot encrypt the keystore: %v", err)
+	}
+
+	f, err := os.OpenFile(temporaryPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, filePermission)
```

Comment:
We should really abstract out on how we write files from the filebeat registry and winlobeat registry: https://github.com/elastic/beats/blob/master/filebeat/registrar/registrar.go#L230 There are strange issues that can happen on shutdown and I remember @andrewkroh fixed on of those in winlogbeat. Having the logic on one place would make sure fixes are applied everywhere. 

Also see the safeFileRotate helper: https://github.com/elastic/beats/blob/master/filebeat/registrar/registrar.go#L251

@ph Could we create a follow up meta issue to track such things we should cleanup after getting this PR in. Happy to help with abstracting out this logic.

Reply:
@ruflin  I have created https://github.com/elastic/beats/issues/5755 for a specific follow up.

Reply:
@ruflin Also create this meta issue https://github.com/elastic/beats/issues/5757 to do a bit more work.